### DISCUSSION TITLE
Support PKCE flow for OAuth apps

### DISCUSSION THREAD

**Discussion by jimafisk** [community/community#15752] [Other Feature Feedback, Questions, & Ideas]
GitHub let's you authorize apps [using OAuth](https://docs.github.com/en/developers/apps/building-oauth-apps/authorizing-oauth-apps) but the supported grants require a server, making it difficult for clientside apps. 

This requires Jamstack git-backed CMS projects like [NetlifyCMS](https://www.netlifycms.org/) manage an intermediary server separately: https://www.netlifycms.org/docs/github-backend/ (same with [TinaCMS](https://tina.io/) and my project [Plenti](https://plenti.co/)).

The [Implicit Grant type](https://datatracker.ietf.org/doc/html/rfc6749#section-4.2) was not allowed because of security issues, but [Proof Key for Code Exchange (PKCE)](https://datatracker.ietf.org/doc/html/rfc7636) should allow clientside apps to operate securely.

GitLab already implements this in their OAuth 2.0 identity provider API: https://docs.gitlab.com/ee/api/oauth2.html#authorization-code-with-proof-key-for-code-exchange-pkce

Thank you!

**Comment by jimafisk** [comment]
Another related discussion: https://github.com/orgs/community/discussions/40077

And documentation issue: https://github.com/github/docs/issues/22270

**Reply by madhurigopal** [reply]
I think we need to Authorize using a local instance of a DB with Cloud connectors something that Langfuse folks do! Do check it out.

**Comment by writeonlycode** [comment]
I would love see support for PKCE as well. It would make a huge difference for JAMStack websites.

**Reply by gopi-suvanam** [reply]
+1

PKCE is important for decoupling services and UI..

**Comment by hickford** [comment]
PR to clarify that PKCE is not supported https://github.com/github/docs/pull/24965

**Comment by Tachi107** [comment]
GitHub now supports the [Device Authorization Grant](https://datatracker.ietf.org/doc/html/rfc8628), which does not require a server.

This is not an excuse to not support PKCE, since its usage is recommended (and sometimes required) starting with the upcoming [OAuth 2.1](https://www.ietf.org/archive/id/draft-ietf-oauth-v2-1-12.html#name-authorization-request) standard.

**Reply by gopi-suvanam** [reply]
In device authorizaiton grant, why is there no `verification_url_complete` ?

**Reply by Evidlo** [reply]
> GitHub now supports the [Device Authorization Grant](https://datatracker.ietf.org/doc/html/rfc8628), which does not require a server.

I think CORS settings on `https://github.com/login/device/code` prevent this.  At least I'm getting CORS errors when I try.

`Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://github.com/login/device/code. (Reason: CORS header ‘Access-Control-Allow-Origin’ missing). Status code: 200.`

https://github.com/octokit/authentication-strategies.js/?tab=readme-ov-file#device-authentication
https://stackoverflow.com/questions/42150075/cors-issue-on-github-oauth

So there is still no way to do client-side auth for GitHub apps!

**Reply by vadimkantorov** [reply]
sigh :(

**Comment by matheuscscp** [comment]
Is PKCE missing only for OAuth Apps or is it missing for GitHub Apps as well? The docs here don't mention "pkce" nor "challenge": https://docs.github.com/en/apps/creating-github-apps/writing-code-for-a-github-app/building-a-login-with-github-button-with-a-github-app

**Reply by Kamil777** [reply]
![Uploading Screenshot_20250715_211955.jpg…]()

**Comment by Tachi107** [comment]
> Is PKCE missing only for OAuth Apps or is it missing for GitHub Apps as well?

As far as I know, it's also missing from GitHub Apps.

**Comment by DDevine** [comment]
I really would prefer if Github had PKCE. It greatly simplifies things for static web applications.

**Comment by hpsin** [comment]
Hey all,
Happy to say we have added support on GitHub.com for PKCE. We'll bring it to GHES in a bit, once we build some tooling for the compat problem

https://github.blog/changelog/2025-07-14-pkce-support-for-oauth-and-github-app-authentication/

**Reply by kyoshino** [reply]
Great news! I’m adding PKCE support to [Sveltia CMS](https://github.com/sveltia/sveltia-cms) now but stuck with a CORS error. Can you please check? (Our PKCE flow with GitLab, Gitea and Forgejo works fine)

> Access to fetch at 'https://github.com/login/oauth/access_token' from origin 'http://localhost:5173' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.

**Reply by kyoshino** [reply]
I have realized that the `client_secret`  param was missing in our requests. It’s not required by [GitLab](https://docs.gitlab.com/api/oauth2/#authorization-code-with-proof-key-for-code-exchange-pkce) or Gitea. (Do we need it??? If so, we still cannot use PKCE for GitHub, because Sveltia CMS runs completely client-side)

But, even when the secret is supplied, requests to `https://github.com/login/oauth/access_token` return a 404 Not Found error with no `Access-Control-Allow-Origin` header. 🤔

**Reply by kyoshino** [reply]
I made it to work. `Content-Type: application/json` or `Accept: application/json` doesn’t work. The CORS mode should be off, and the body must be `application/x-www-form-urlencoded`.

```js
await fetch('https://github.com/login/oauth/access_token', {
  method: 'POST',
  mode: 'no-cors',
  body: new URLSearchParams({ client_secret, client_id, code, redirect_uri, code_verifier }),
});
```

But, omitting `client_secret` results in an error. So unfortunately we cannot use this method. 😢

**Reply by hpsin** [reply]
We still have to add support for CORS on the token endpoint unfortunately. We are working on it though.

For the client secret, yes, our implementation always requires the client secret as we don't yet distinguish between public clients and confidential clients. So at this time, native apps and SPA type apps need to include their client secret in the redemption. I don't have an estimate for when we can remove this requirement unfortunately.

**Reply by kyoshino** [reply]
We cannot embed a client secret in our SPA or ask end-users to enter their client secret. So we have to wait until the requirement is removed. We could rather prompt users to enter their personal access token directly, as discussed here:

- https://github.com/decaporg/decap-cms/issues/663

**Reply by Evidlo** [reply]
So just to confirm, we still cannot authenticate client-side webapps without a backend with Github?

It's currently possible to [generate a link](https://github.com/settings/tokens/new?description=foobarapp&scopes=gist&default_expires_at=none) to the personal access token page which could be copy-pasted into the app to give it access to the user's repositories/gists/etc, but this is not very ergonomic.

**Reply by hpsin** [reply]
Ah, I'm sorry to hear that. Correct, you cannot yet auth a SPA as we don't have CORS for the token endpoint yet. We'd really love to not have folks create PATs for this, so I'll update when I have more details about the CORS options here.

**Reply by kyoshino** [reply]
Once the CORS issue is resolved, I would try the [device flow](https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/authorizing-oauth-apps#device-flow) as [discussed earlier](https://github.com/orgs/community/discussions/15752#discussioncomment-11536715), which doesn’t require the `client_secret` param.

Our situation is a bit complicated. [Sveltia CMS](https://github.com/sveltia/sveltia-cms) is not SaaS but (semi) self-hosted. We serve the SPA via UNPKG, and users have their own instances. Since the `redirect_uri` varies by user, each user must install an [OAuth client](https://github.com/sveltia/sveltia-cms-auth) on Cloudflare Workers and register it with GitHub, which is cumbersome. From a UX standpoint, prompting for a PAT is much simpler.

Anyway, it’s great to hear that GitHub is working on this. Keep it going!

**Reply by vadimkantorov** [reply]
@kyoshino If we self-build Sveltia CMS (to customize redirect_uri) and serve it from the end-user domain, would this new flow allow using GitHub API auth in a backend-less manner?

@hpsin This flow should enable completely server-less CMS for GitHub repo-backed sites with proper login forms and identity management

**Reply by kyoshino** [reply]
@vadimkantorov Sveltia’s PAT sign-in option doesn’t rely on the PKCE thing. You don’t need to build the CMS yourself. Simply enter your PAT and subsequent API requests will be made directly between the CMS and GitHub.

**Reply by vadimkantorov** [reply]
My question was: if you build sveltia yourself and burn in custom http return url, can you use the PKCE thing instead of PAT?

**Reply by kyoshino** [reply]
The [redirect URL](https://docs.github.com/en/apps/creating-github-apps/registering-a-github-app/about-the-user-authorization-callback-url) is specified in the application settings on the GitHub web UI, rather than being embedded in the CMS.

You still cannot use PKCE until GitHub supports CORS and makes the client secret optional.

**Reply by vadimkantorov** [reply]
Oh, got it :( is there a technically possible option for the user to create their own github app / install it in their CMS repo and specify there a custom return url?

**Reply by kyoshino** [reply]
You can (and you need to) create you own app to specify a custom return URL.

**Reply by vadimkantorov** [reply]
Got it. And does Sveltia also support this mode? This (admittedly cumbersome) mode can be a useful sometimes for multi-contributor websites (while waiting for proper no-backend mode)

**Reply by kyoshino** [reply]
It’s an OAuth (or GitHub) application setting, not something that CMS frontend has to support. See https://github.com/sveltia/sveltia-cms-auth

**Reply by vadimkantorov** [reply]
Oh, I see - this return url must be a backend-ful app? Could it be just an HTML static page? (e.g. hosted on the website's domain static hosting)

Meaning: could the redirect be handled by pure JavaScript code embedded in HTML / index.html?

**Reply by seanorama** [reply]
Any word on when this will land in GHES?

**Reply by imjasonh** [reply]
@hpsin Any update on CORS support?

Is there a separate issue/discussion I can subscribe to for that?

**Comment by gopi-suvanam** [comment]
Is it safe to keep client secret in the front end.. it won't be a "secret"
any more

**Reply by hpsin** [reply]
Yes, we don't have a "public client" concept yet, so we treat all clients the same and all of them require access to the client secret. You cannot keep a secret "secret" within a public client, but you do have to embed it there anyhow. This is how e.g. VS Code, Visual Studio, GH CLI, and GitHub Mobile all work. 

You can read a bit more about that here: https://docs.github.com/en/enterprise-cloud@latest/apps/creating-github-apps/about-creating-github-apps/best-practices-for-creating-a-github-app#client-secrets

**Reply by monarchwadia** [reply]
@hpsin I am building a desktop app. I want to allow the user to connect to Github and provide my app with read access. 

My questions are:

1. My understanding is that I will have to ship the client_secret, thus making it a "public client secret"... is that a fair assessment? Is this Github's recommended practice, given that VSCode, Visual Studio, GH CLI, Github Mobile all do this?
2. What are the right steps for mitigating risk on my end? 
3. And, also, how does Github mitigate risk for this use case on its end? What is the security model like?

**Reply by hpsin** [reply]
1. Yes, that's the case. We want to move away from that since it's a common thing people trip up on. This is on the public roadmap here: https://github.com/github/roadmap/issues/1153 (native apps to come, but a similar outcome) 
2. Ensure that you are not making assumptions on a backend that a token from "your app" (meaning from your client ID) is de facto trustable. (I.e. "This token came from XX App so it can call this API"). This is true regardless of whether you have a client secret or not because you cannot validate the legitimacy of a public client. Beyond that there's nothing you need to do. The client secret can only be used to manipulate a token someone has in possession - there's nothing it can do without a token. 
3. As above :)

**Reply by monarchwadia** [reply]
Thank you @hpsin that makes sense. I appreciate the response!

**Comment by gopi-suvanam** [comment]
We are also Asking user to enter the token. Will wait for fully client side
authentication to be enabled

**Comment by mig5** [comment]
I am using Vouch which was already sending PKCE code_challenge_method S256 to Github with a correct 43 char length of code_challenge, yet I got the email saying my Github OAuth app was using PKCE incorrectly..

I set up a new Github OAuth app and it Just Worked despite me not changing anything other than the client ID and client secret in my Vouch config. (I also set the `redirect_uri` which is optional but which Github 'strongly recommends' [according to the docs](https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/authorizing-oauth-apps#2-users-are-redirected-back-to-your-site-by-github)

However, after it working the first time, I now get a 400 error with the Github side returning: ` oauth2: \"bad_verification_code\" \"The code passed is incorrect or expired.\" \"https://docs.github.com/apps/managing-oauth-apps/troubleshooting-oauth-app-access-token-request-errors/#bad-verification-code\""}`

This is weird, it was working ok on the older OAuth apps...? So I'm doubly confused why the old apps were apparently 'incorrect', and why the new app worked only once, despite me not having changed other parameters.

I'm pretty sure Vouch does everything correctly, it's a solid OIDC sidecar for Nginx.

I've filed an upstream issue at https://github.com/vouch/vouch-proxy/issues/600 about it though.

**Comment by kyoshino** [comment]
@hpsin It’s been six months since GitHub missed its Q4 2025 goal to release [PKCE support for SPA](https://github.com/github/roadmap/issues/1153). Any news about this?

Our [Sveltia CMS](https://sveltiacms.app/en/), the de facto successor to Netlify CMS mentioned in the OP, is becoming more popular. We estimate that it now has thousands of users worldwide.

Sveltia CMS has supported PAT authentication since July 2025, but the number of installations of [our own OAuth client](https://github.com/sveltia/sveltia-cms-auth), Sveltia CMS Authenticator, [continues to increase](https://github.com/search?q=sveltia-cms-auth&type=repositories&s=updated&o=desc) due to the lack of PKCE support.

We really hope the feature will be implemented soon.

**Comment by MoritzWeber0** [comment]
Is there any addition information on why [#1153](https://github.com/github/roadmap/issues/1153) was changed to paused?
